Fix CakeSocket not being able to connect to TLS1.2 only servers

markstory opened the pull request:
Include the changes from #10232 and also fix the issue described in #10445

mstramek-CKPL committed:
#10230 - Extend available TLS encrypt methods in CakeSocket

mstramek-CKPL committed:
#10230 - Extend available TLS encrypt methods in CakeSocket - formatting

markstory committed:
Conditionally define TLS constants & add test for new types.

Only map the new TLS constants if they exist.

markstory committed:
Update tls_client and tls_server to include newer TLS versions.

Refs #10445

ravage84 commented on lib/Cake/Network/CakeSocket.php:
```diff
@@ -116,6 +117,32 @@ class CakeSocket {
+		// These TLS versions are not supported by older PHP versions,
+		// so we have to conditionally set them if they are supported.
+		$conditionalCrypto = array(
+			'tlsv1_1_client' => 'STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT',
```
If we set these TLS versions depending on the PHP version, shouldn't we test the list of them conditionally?

espaistelematics replied:
Perfect! Works good.

ravage84 commented on lib/Cake/Network/CakeSocket.php:
```diff
@@ -116,6 +117,32 @@ class CakeSocket {
 	public function __construct($config = array()) {
 		$this->config = array_merge($this->_baseConfig, $config);
+
+		// These TLS versions are not supported by older PHP versions,
```
The whole TLS related code could be moved into a protected method in which the doc blcok could epxlain the situation.

 + smaller method
 + related stuff is extracted and contained
 + in line comment moves to class API  docs

markstory replied:
good idea

ravage84 commented on lib/Cake/Network/CakeSocket.php:
```diff
@@ -44,7 +44,8 @@ class CakeSocket {
 		'port' => 80,
-		'timeout' => 30
+		'timeout' => 30,
+		'cryptoType' => 'tls'
```
Nitpicking: trailing comma.

ravage84 reviewed (commented):
Looks good.

RichieB2B commented on lib/Cake/Network/CakeSocket.php:
```diff
@@ -116,6 +117,32 @@ class CakeSocket {
+		// These TLS versions are not supported by older PHP versions,
+		// so we have to conditionally set them if they are supported.
+		$conditionalCrypto = array(
+			'tlsv1_1_client' => 'STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT',
```
This is done by the `if (defined($const))` on line 130. If PHP doesn't know about the constant, it won't get added.

markstory committed:
Split conditional TLS versions into a separate method

Having a separate method gives a nicer home to the inline comments. I've
had to disable coding standards checks as the constants provided by PHP
do not follow the UPPER_CASE conventions.

markstory committed:
Fix failing test.

This test merges config funny.
